File: deviceManager/Tcp_client_manager.py
import socket
import json
from datetime import datetime
import struct
import time

class Tcp_client_manager():
    def __init__(self):
        self.TCP_HOST = "192.168.0.86"
        # self.TCP_HOST = "172.20.10.8"
        self.TCP_PORT = 1201
        self.PATROL_NUMBER = 1
        self.deviceManager_ID = 0x01
        self.situationDetector_ID = 0x02
        self.recv_data = ""
        self.alarm = 0
        self.last_alarm = 0
        
    def socket_init(self):
        self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        connected = False
        while (not connected):
            try:
                print("Socket Waiting for Connection from situationDetector ")
                self.client_socket.connect((self.TCP_HOST, self.TCP_PORT))
                connected = True
                print("situationDetector Connected")
            except Exception as e:
                print(f"Socket Initiate Error Occured : {e}")
                time.sleep(5)

    def send_data(self, data):
        self.client_socket.send(data)

    def send_data_all(self, data):
        self.client_socket.sendall(data)
        
    def receive_data(self):
        data_size = 4
        while True:
            try:
                self.client_socket.settimeout(2.0)
                self.recv_data = self.client_socket.recv(data_size)
                self.data_validation()
                time.sleep(0.1)
            except Exception:
                pass    
            except KeyboardInterrupt:
                print("Keyboard Interruption")
                break
    
    def data_validation(self):
        if (self.recv_data[0] != self.situationDetector_ID):
            print("Wrong Data Source Input")
            self.recv_data = ""
            return
        if (self.recv_data[1] != self.deviceManager_ID):
            print("Wrong Data Destination Input")
            self.recv_data = ""
            return
        if (self.recv_data[2] != self.PATROL_NUMBER):
            print("Wrong Patrol Number")
            self.recv_data = ""
            return
        self.alarm = int(self.recv_data[3])
        
        if (self.alarm != 0):
            self.last_alarm = self.alarm


    def socket_close(self):
        self.client_socket.close()

    def __exit__(self):
        print("tcp close")
        self.socket_close()

File: deviceManager/TCPClientManager.cpp
#include "TCPClientManager.hpp"



TCPClientManager::TCPClientManager(char* host, int port, u_int8_t patrol_number, u_int8_t device_id, u_int8_t detector_id)
{
    PATROL_NUMBER = patrol_number;
    deviceManager_ID = device_id;
    situationDetector_ID = detector_id;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    server_addr.sin_addr.s_addr = inet_addr(host);
    recv_data.store({0, 0, 0, 0});
}

TCPClientManager::~TCPClientManager()
{
    socket_close();
}


void TCPClientManager::socket_init()
{
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if(client_socket < 0){
        cerr << "Socket creation error \n";
        exit(EXIT_FAILURE);
    }

    while(true){
        if(connect(client_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) == 0){
        cout << "situationDetector Connected \n";
        break;
        }
        else{
            cerr << "Connection Failed \n";
            this_thread::sleep_for(std::chrono::seconds(3));
        }
    }
}

void TCPClientManager::socket_close()
{
    if (client_socket >= 0){
        close(client_socket);
        client_socket = -1;
    }
    
}

void TCPClientManager::send_data(uint8_t *data, size_t length)
{
    send(client_socket, data, length, 0);
}

void TCPClientManager::receive_data()
{
    while(true){
        int valread = recv(client_socket, &recv_data, sizeof(recv_data), 0);
        if(valread > 0){
            date_validation();
        }
        
        this_thread::sleep_for(chrono::milliseconds(100));
    }   
}

void TCPClientManager::date_validation()
{
    if (recv_data.load().source == 0x02 && recv_data.load().destination == deviceManager_ID && recv_data.load().patrol_number == PATROL_NUMBER){
        // cout << "Valid Data Received" << endl;
    }
    else{
        cout << "Invalid Data Received" << endl;
        recv_data.store({0, 0, 0, 0});
    }
}

uint8_t TCPClientManager::load_data()
{
    return recv_data.load().alarm_type;
}

void TCPClientManager::save_data(uint8_t data)
{
    recv_data.store({deviceManager_ID, situationDetector_ID, PATROL_NUMBER, data});
}
